internal/daemon/controller/handlers/targets/target_service.go

```diff
@@ -234,87 +234,115 @@ func (s Service) ListTargets(ctx context.Context, req *pbs.ListTargetsRequest) (
-		opts := []target.Option{
-			target.WithLimit(limit),
+	filterItemFn := func(item target.Target) (bool, error) {
+		pbItem, err := toProto(ctx, item, newOutputOpts(ctx, item, authResults, authzScopes)...)
+		if err != nil {
+			return false, err
 		}
-		if prevPageLast == nil {
-			if refreshToken != nil {
-				opts = append(opts, target.WithStartPageAfterItem(
-					refreshToken.GetLastItemId(),
-					refreshToken.GetLastItemUpdatedTime().AsTime(),
-				))
-			}
-		} else {
-			opts = append(opts, target.WithStartPageAfterItem(prevPageLast.GetPublicId(), prevPageLast.GetUpdateTime().AsTime()))
+
+		filterable, err := subtypes.Filterable(pbItem)
+		if err != nil {
+			return false, err
 		}
-		return repo.ListTargets(ctx, opts...)
+		return filter.Match(filterable), nil
 	}
-	filterAndConvertFn := func(item target.Target) (*pb.Target, error) {
-		pr := perms.Resource{Id: item.GetPublicId(), ScopeId: item.GetProjectId(), Type: resource.Target}
-		outputFields := authResults.FetchOutputFields(pr, action.List).SelfOrDefaults(authResults.UserId)
 
-		outputOpts := make([]handlers.Option, 0, 3)
-		outputOpts = append(outputOpts, handlers.WithOutputFields(outputFields))
-
-		if outputFields.Has(globals.ScopeField) {
-			outputOpts = append(outputOpts, handlers.WithScope(authzScopes[item.GetProjectId()]))
-		}
-		if outputFields.Has(globals.AuthorizedActionsField) {
-			authorizedActions := authResults.FetchActionSetForId(ctx, item.GetPublicId(), IdActions, auth.WithResource(&pr)).Strings()
-			outputOpts = append(outputOpts, handlers.WithAuthorizedActions(authorizedActions))
+	var listResp *target.ListResponse
+	if req.GetRefreshToken() == "" {
+		listResp, err = target.List(ctx, repo, &authResults, pageSize, filterItemFn)
+		if err != nil {
+			return nil, err
 		}
-
-		pbItem, err := toProto(ctx, item, outputOpts...)
+	} else {
+		rt, err := parseRefreshToken(ctx, req.GetRefreshToken())
 		if err != nil {
 			return nil, err
 		}
-
-		filterable, err := subtypes.Filterable(pbItem)
+		var domainResourceType refreshtoken.ResourceType
+		switch rt.ResourceType {
+		case pbs.ResourceType_RESOURCE_TYPE_TARGET:
+			domainResourceType = refreshtoken.ResourceTypeTarget
+		// TODO: Add more when generalizing this
+		default:
+			domainResourceType = refreshtoken.ResourceTypeUnknown
+		}
+		// We're doing the conversion from the protobuf types to the
+		// domain types here rather than in the domain so that the domain
+		// doesn't need to know about the protobuf types.
+		domainRefreshToken := &refreshtoken.RefreshToken{
+			CreatedTime:         rt.CreatedTime.AsTime(),
+			ResourceType:        domainResourceType,
+			PermissionsHash:     rt.PermissionsHash,
+			LastItemId:          rt.LastItemId,
+			LastItemUpdatedTime: rt.LastItemUpdatedTime.AsTime(),
+		}
+		err = refreshtoken.ValidateRefreshToken(
+			ctx,
+			domainRefreshToken,
+			refreshtoken.ResourceTypeTarget,
+			&authResults,
+		)
```

Comment:
The refresh token is validated in the application layer as it is part of request validation, but this could be moved to the domain if preferred.